### PR TITLE
[DropIn] Remove back press functions from DropInNavigationView

### PR DESCRIPTION
**Pull request by kmadsen**
### Description

Similar to https://github.com/mapbox/mapbox-navigation-android/pull/5585
Resolves [#1513](https://github.com/mapbox/navigation-sdks/issues/1513)

Right now you must hook up the `DropInNavigationView` to your `ComponentActivity` using the `getOnBackPressedCallback` function. This breaks our rule for "zero setup" with the drop in view. So this pull request proposes we try `setOnKeyListener`.

We may have an issue with `setOnKeyListener`, considering it [differentiates between hardware and software keys](https://developer.android.com/reference/android/view/View.OnKeyListener). We have not found a device where this is an issue yet.

### Alternative: If there is an issue with `setOnKeyListener`

Another approach discussed is to add another `unwrapContext` in the DropInNavigationView to find the `onBackPressedDispatcher` in a hosting `ComponentActivity`. The concern here is that we require a ComponentActivity to host the `DropInNavigationView`. This may be okay, but we are trying the "view first api".
``` kotlin
private fun Context.toOnBackPressedDispatcher(): OnBackPressedDispatcher {
    val componentActivity = recursiveUnwrap(this) as? ComponentActivity
    checkNotNull(componentActivity) {
        "Please ensure that the hosting Context is a valid ComponentActivity"
    }
    return componentActivity.onBackPressedDispatcher
}
```

### Alternative: We could also override `onKeyUp`

I think this will run into the same issues as `setOnKeyListener` because it the underlying mechanisms are the same at first glance. Am proposing we use  `setOnKeyListener` inside a lifecycle component so it is easy to "detach" at runtime. The `onKeyUp` can also be disabled.. 

The tie-breaker is to keep logic outside of `DropInNavigationView` and use a lifecycle aware `UIComponent`.

#### Alternative: Go to the future

Looks like Android 13 is adding better support for handling back pressed with https://developer.android.com/reference/android/view/OnBackInvokedCallback#onBackInvoked()

### CODE REVIEW

**Comment by kmadsen** [comment]
The old `NavigationView` required back press to be forwarded in the same way we have on the main line

https://github.com/mapbox/mapbox-navigation-android/blob/e232efd36700e4f5f3345bd76ad99cb28076bb1e/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java#L129-L137

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/backpress/OnKeyListenerComponent.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            }
+        }
+    }
+
+    private fun handleBackPress(): Boolean {
+        return when (navigationStateViewModel.state.value) {
+            NavigationState.FreeDrive, NavigationState.Empty -> {
+                // Do not handle the back press in FreeDrive without a destination. FreeDrive is
+                // the initial navigation state, returning false will give control to the hosting
+                // back stack gesture detector.
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/backpress/OnKeyListenerComponent.kt:48]: This comment is probably not relevant anymore @kmadsen

File: libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/component/backpress/OnKeyListenerComponentTest.kt [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        val mockMapboxNavigation: MapboxNavigation = mockk()
+        onKeyListenerComponent.onAttached(mockMapboxNavigation)
+        onKeyListenerComponent.onDetached(mockMapboxNavigation)
+
+        verify { view.isFocusableInTouchMode = false }
+        verify { view.setOnKeyListener(null) }
+    }
+
+    @Test
+    fun `pack press is not handled when in FreeDrive state without destination`() {
```
Comment (abhishek1508) [inline libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/component/backpress/OnKeyListenerComponentTest.kt:63]:
```suggestion
    fun `back press is not handled when in FreeDrive state without destination`() {
```

File: libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/component/backpress/OnKeyListenerComponentTest.kt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        )
+
+        onKeyListenerComponent.onAttached(mockk())
+        val result = slotOnKeyListener.captured.triggerBackPressed()
+
+        assertFalse(result)
+    }
+
+    @Test
+    fun `pack press will set destination to null when in FreeDrive with a destination`() {
```
Comment (abhishek1508) [inline libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/component/backpress/OnKeyListenerComponentTest.kt:78]:
```suggestion
    fun `back press will set destination to null when in FreeDrive with a destination`() {
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/backpress/OnKeyListenerComponent.kt (hunk 1 above)
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/backpress/OnKeyListenerComponent.kt:48]: yeah you're right, tests also help with understanding

File: libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/component/backpress/OnKeyListenerComponentTest.kt (hunk 2 above)
Comment (kmadsen) [reply libnavui-dropin/src/test/java/com/mapbox/navigation/dropin/component/backpress/OnKeyListenerComponentTest.kt]: haha yeah all of them say pack

fixed

**Review by abhishek1508 — APPROVED** [review]
Based on this PR, when in active navigation state and back button is pressed, we stop navigation and go back to FreeDrive. We should have some discussion on the whether we would want to go back to FreeDrive or RoutePreview upon back press on Active Nav. This is not a blocker for this PR.

**Comment by kmadsen** [comment]
> Based on this PR, when in active navigation state and back button is pressed, we stop navigation and go back to FreeDrive. We should have some discussion on the whether we would want to go back to FreeDrive or RoutePreview upon back press on Active Nav. This is not a blocker for this PR.

Yeah there are issues with determining RoutePreview that are out of scope for this PR. Starting to address more of those issues here https://github.com/mapbox/mapbox-navigation-android/pull/5594

**Pull request merged**